Add per_element_output_counter to python SDK

boyuanzz opened the pull request:
Add per_element_output_counter to python worker.
CounterName: per-element-output-count + operation_name
Experimental flag: outputs_per_element_counter

Related PR: https://github.com/apache/beam/pull/4613
r: @pabloem  @charlesccychen

boyuanzz commented:
Hi @charlesccychen , could you please review my code changes when you have time? Thanks for your help!

robertwb commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -535,6 +545,17 @@ def __init__(self, window_fn, main_receivers, tagged_receivers):
+      self.per_element_output_counter = counter_factory.get_counter(
+          per_element_output_counter_name, Counter.DISTRIBUTION)
+    else:
+      self.per_element_output_counter = counter_factory.get_counter(
```
If the experiment is disabled, we're still doing all the work, but just not publishing the result?

boyuanzz replied:
I'll change this part to do nothing.

robertwb commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -570,6 +594,7 @@ def process_outputs(self, windowed_input_element, results):
         self.main_receivers.receive(windowed_value)
       else:
         self.tagged_receivers[tag].receive(windowed_value)
+    self.per_element_output_counter.update(output_element_output)
```
Anything we do per-element is very performance sensitive. Profiles?

boyuanzz replied:
I ran wordcount pipeline with 1gb input 10 times both with and without per_element_output_counter. I got result as following:
+------------|  with     | without |
avg(runtime) | 758.5s |  681s .   |
+------------------------------+

robertwb commented:
I ran some benchmarks, this PR more than doubles the overhead of handling an element (i.e. updating per_element_output_counter costs more than everything else we do in this code). 

```
import time
from scipy import stats
import apache_beam as beam

num_maps = 100
timings = {}
for run in range(5):
  num_elements = 1000 * run + 1
  start = time.time()
  with beam.Pipeline() as p:
    pc = p | beam.Create(range(num_elements))
    for ix in range(num_maps):
      pc = pc | 'Map%d' % ix >> beam.FlatMap(lambda x: (None,))
  timings[num_elements] = time.time() - start
  print sorted(timings.items())

gradient,intercept,r_value,p_value,std_err=stats.linregress(*zip(*timings.items()))
print "Fixed cost ", intercept
print "Per-element", gradient / num_maps
print "R^2        ", r_value**2
```

With this change: 

```
[(1, 1.539193868637085)]
[(1, 1.539193868637085), (1001, 1.670807123184204)]
[(1, 1.539193868637085), (1001, 1.670807123184204), (2001, 2.1758179664611816)]
[(1, 1.539193868637085), (1001, 1.670807123184204), (2001, 2.1758179664611816), (3001, 2.417663812637329)]
[(1, 1.539193868637085), (1001, 1.670807123184204), (2001, 2.1758179664611816), (3001, 2.417663812637329), (4001, 2.832465171813965)]
Fixed cost  1.46017638946
Per-element 3.33339929581e-06
R^2         0.976258118233
```

Commenting out the three new lines in process_outputs

```
[(1, 1.46219801902771)]
[(1, 1.46219801902771), (1001, 1.4264578819274902)]
[(1, 1.46219801902771), (1001, 1.4264578819274902), (2001, 1.735224962234497)]
[(1, 1.46219801902771), (1001, 1.4264578819274902), (2001, 1.735224962234497), (3001, 1.790186882019043)]
[(1, 1.46219801902771), (1001, 1.4264578819274902), (2001, 1.735224962234497), (3001, 1.790186882019043), (4001, 2.024906873703003)]
Fixed cost  1.38981666722
Per-element 1.48914670944e-06
R^2         0.90302770475
```

(This is without enabling the experiment.)

boyuanzz commented:
Things changed:
1. Add DistributionMetadataCounter, append per_element_output to counter's list during processing_output
2. Will update original DistributionCounter within progress_report thread

Performance improved(wordcount pipeline with 1gb input):
New AVG(runtimes) = 703s
Old AVG(runtimes) = 758s
Without counter = 681s

@robertwb  Please take a look at new changes, thanks :D

charlesccychen commented:
Can you run the overhead benchmarks with the code Robert posted above?

boyuanzz commented:
Results ran on my machine:
With Changes:
```
[(1, 0.9140629768371582)]
[(1, 0.9140629768371582), (1001, 0.9695911407470703)]
[(1, 0.9140629768371582), (1001, 0.9695911407470703), (2001, 1.1344990730285645)]
[(1, 0.9140629768371582), (1001, 0.9695911407470703), (2001, 1.1344990730285645), (3001, 1.1834659576416016)]
[(1, 0.9140629768371582), (1001, 0.9695911407470703), (2001, 1.1344990730285645), (3001, 1.1834659576416016), (4001, 1.3307790756225586)]
Fixed cost  0.896913511181
Per-element 1.04730701447e-06
R^2         0.972517109068
```
WIthout Changs:
```
[(1, 0.8925869464874268)]
[(1, 0.8925869464874268), (1001, 1.0654640197753906)]
[(1, 0.8925869464874268), (1001, 1.0654640197753906), (2001, 1.1542890071868896)]
[(1, 0.8925869464874268), (1001, 1.0654640197753906), (2001, 1.1542890071868896), (3001, 1.2227840423583984)]
[(1, 0.8925869464874268), (1001, 1.0654640197753906), (2001, 1.1542890071868896), (3001, 1.2227840423583984), (4001, 1.2981250286102295)]
Fixed cost  0.932873731899
Per-element 9.68396186829e-07
R^2         0.957529775313
```
@robertwb  @charlesccychen

boyuanzz commented:
jenkins retest this please

boyuanzz commented:
I ran this script again under Cython env:
With counter:
```
[(1, 0.9012792110443115)]
[(1, 0.9012792110443115), (1001, 0.9603300094604492)]
[(1, 0.9012792110443115), (1001, 0.9603300094604492), (2001, 1.1487770080566406)]
[(1, 0.9012792110443115), (1001, 0.9603300094604492), (2001, 1.1487770080566406), (3001, 1.2948799133300781)]
[(1, 0.9012792110443115), (1001, 0.9603300094604492), (2001, 1.1487770080566406), (3001, 1.2948799133300781), (4001, 1.3819458484649658)]
Fixed cost  0.878136174011
Per-element 1.29588317871e-06
R^2         0.97726061306
```
Without counter:
```
[(1, 0.897953987121582)]
[(1, 0.897953987121582), (1001, 0.9599549770355225)]
[(1, 0.897953987121582), (1001, 0.9599549770355225), (2001, 1.2793560028076172)]
[(1, 0.897953987121582), (1001, 0.9599549770355225), (2001, 1.2793560028076172), (3001, 1.1948518753051758)]
[(1, 0.897953987121582), (1001, 0.9599549770355225), (2001, 1.2793560028076172), (3001, 1.1948518753051758), (4001, 1.2624180316925049)]
Fixed cost  0.926045594811
Per-element 9.63824987411e-07
R^2         0.7361417237
```
@robertwb  @charlesccychen

boyuanzz commented:
Hi @robertwb and @charlesccychen , could you please review my changes and performance? Please feel free to ping me if you have any concerns. Thanks for you all help!

charlesccychen commented:
I'm concerned that we may need to do more optimization, since a 30% regression on this performance-critical code path might not be something we can accept; however I will defer to @robertwb on this.

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -594,7 +596,9 @@ def process_outputs(self, windowed_input_element, results):
-    self.per_element_output_counter.update(output_element_output)
+    # TODO: remove if block after per-element-output-counter released
+    if self.has_per_element_output_count:
+      self.per_element_output_counter.update(output_element_output)
```
What I think Robert means is to use `@cython.locals` to make sure the added variable is typed as an int when translated by Cython into C++.  See for example:

https://github.com/boyuanzz/beam/blob/7cf9c0b16b4fe15b0eab3f73860fc560ceebff67/sdks/python/apache_beam/runners/common.pxd#L89

boyuanzz replied:
Fixed in commit https://github.com/apache/beam/pull/4741/commits/6c8d60bbfb2732c20d13ce00cddb83e7be73cd40

boyuanzz commented:
Hi @robertwb , @charlesccychen , @swegner ,
I ran Robert's script against both origin:master and my feature branch(without counter enabled).
origin:master:
```
[(1, 0.9149210453033447)]
[(1, 0.9149210453033447), (1001, 0.9606211185455322)]
[(1, 0.9149210453033447), (1001, 0.9606211185455322), (2001, 0.9848501682281494)]
[(1, 0.9149210453033447), (1001, 0.9606211185455322), (2001, 0.9848501682281494), (3001, 1.1837430000305176)]
[(1, 0.9149210453033447), (1001, 0.9606211185455322), (2001, 0.9848501682281494), (3001, 1.1837430000305176), (4001, 1.2574291229248047)]
Fixed cost  0.878594469857
Per-element 9.08138036728e-07
R^2         0.90765971031
```
feature branch without counter enabled:
```
[(1, 0.9121770858764648)]
[(1, 0.9121770858764648), (1001, 0.9933300018310547)]
[(1, 0.9121770858764648), (1001, 0.9933300018310547), (2001, 0.9446542263031006)]
[(1, 0.9121770858764648), (1001, 0.9933300018310547), (2001, 0.9446542263031006), (3001, 1.1995000839233398)]
[(1, 0.9121770858764648), (1001, 0.9933300018310547), (2001, 0.9446542263031006), (3001, 1.1995000839233398), (4001, 1.291064977645874)]
Fixed cost  0.875259707403
Per-element 9.63945865631e-07
R^2         0.828742120861

```

charlesccychen commented on sdks/python/apache_beam/runners/common.pxd:
```diff
@@ -86,15 +86,16 @@ cdef class DoFnRunner(Receiver):
 
 cdef class OutputProcessor(object):
-  @cython.locals(windowed_value=WindowedValue)
+  @cython.locals(windowed_value=WindowedValue, output_element_output=cython.int)
```
<!--new_thread; commit:6c8d60bbfb2732c20d13ce00cddb83e7be73cd40; resolved:0-->
You can do `cdef bint has_per_element_output_count` here.

boyuanzz replied:
I'm not sure whether I understand this comment correctly, but I defined 'has_per_element_output_count' in 'cdef class _OutputProcessor', which is a derived class of OutputProcessor

charlesccychen replied:
Ah, my mistake.

boyuanzz commented:
Benchmark tests twice with feature disabled(1gb input wordcount pipeline, 10 times each):
AVG(runtime) of 1st run: 687s
AVG(runtime) of 2nd runt: 679s

Compared to master sdk:
AVG(runtime) = 681s

@robertwb  @charlesccychen @swegner

swegner commented:
Thanks for providing timing results @boyuanzz. Because this shows no performance regression in the default state (experiment disabled), I recommend we file a JIRA to come back and tune performance before switching it on by default. Getting this merged will allow us to more easily evaluate it against existing benchmarks. @robertwb what do you think?

We also need to decide on what level of performance overhead is acceptable for adding such instrumentation. I propose we choose an existing benchmark which resembles a real user pipeline, and evaluate the delta for total wall time.

charlesccychen commented:
A less noisy and more direct way of measuring the overhead is with the benchmark script Robert posted above, which stresses the part of execution this affects.  End-to-end runtime is too noisy since it may include variable GCE VM startup / autoscaling variations.  I ran the script with and without this change, averaging over 10 runs in each scenario, and it shows a +1.16% change in per-element overhead (1.443362569807e-06s with this change vs 1.426719498634e-06 before), when the flag is turned off.  Robert, what do you think?

robertwb commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -536,16 +546,32 @@ def __init__(self, window_fn, main_receivers, tagged_receivers):
     self.tagged_receivers = tagged_receivers
 
+    # TODO: remove if block after per-element-output-counter released
+    experiments = RuntimeValueProvider.get_value('experiments', str, '')
```
File a JIRA about profiling this before enabling, and reference here.

boyuanzz replied:
Will add a TODO here in next revision

robertwb commented on sdks/python/apache_beam/runners/direct/evaluation_context.py:
```diff
@@ -224,7 +225,10 @@ def handle_result(
                 self._side_inputs_container.finalize_value_and_get_tasks(view))
 
       if result.counters:
+        # there is no merge fn in a distribution counter
```
Shouldn't there be?

boyuanzz replied:
For current purpose, we only update counter and translate counter into update_metric_proto. How about add 'pass' to distribution_counter.merge func?

robertwb replied:
What I don't understand is why we have a special case to skip it in this case (and not others). 

If we want a dummy implementation, raising NotImplemntedError is preferable to a pass in case someone tries to call it.

boyuanzz replied:
Right, I've made some changes in latest revision. Actually, I'll translate DistributionMetadataAccumulator into DistributionAccumulator in batchworker. So for a counter_factory in one work_item, DistributionMetadataAccumulator is the only one new counter I added. I'll raise NotImplemntedError in next commit. Thanks!

boyuanzz commented:
jenkins retest this please

boyuanzz commented:
Hey @robertwb , please take a look at latest changes, which addressed your last comments. Thanks :P

boyuanzz commented:
Hey @charlesccychen ,
I reran robertwb's script 10 times with and without counter to get a average per-element overhead:
```
With counter enabled: AVG(per-element) = 9.37827587128e-07
Master branch: AVG(per-element) = 8.78213119507e-07
```
It's like aggression will no more than 6% per element

boyuanzz commented:
Hey @charlesccychen,

Please take a look at this PR. All comments above are addressed and python-related tests passed. Thanks!

charlesccychen commented on sdks/python/apache_beam/runners/dataflow/internal/apiclient.py:
```diff
@@ -744,6 +744,12 @@ def translate_distribution(distribution_update, metric_update_proto):
   dist_update_proto.max = to_split_int(distribution_update.max)
   dist_update_proto.count = to_split_int(distribution_update.count)
   dist_update_proto.sum = to_split_int(distribution_update.sum)
+  if hasattr(distribution_update, 'buckets') and \
```
Please avoid line continuation (wrap everything in `(...)`).

boyuanzz replied:
will fix it in next revision.

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -592,6 +619,9 @@ def process_outputs(self, windowed_input_element, results):
         self.main_receivers.receive(windowed_value)
       else:
         self.tagged_receivers[tag].receive(windowed_value)
+    # TODO: remove if block after per-element-output-counter released
```
Please reference a JIRA ticket.

boyuanzz replied:
I'll attach one here. Thanks!

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -558,16 +568,33 @@ def __init__(self, window_fn, main_receivers, tagged_receivers):
+    experiments = RuntimeValueProvider.get_value('experiments', str, [])
+    if 'outputs_per_element_counter' in experiments:
+      self.has_per_element_output_count = True
+      per_element_output_counter_name = \
```
Please avoid \ line continuations.  Instead, do

```python
per_element_output_counter_name = (
    CounterName('per-element-output-count-meta', step_name=operation_name))
```

boyuanzz replied:
will fix in next revision.

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -558,16 +568,33 @@ def __init__(self, window_fn, main_receivers, tagged_receivers):
     self.main_receivers = main_receivers
     self.tagged_receivers = tagged_receivers
 
+    # TODO: remove if block after per-element-output-counter released
```
Please file JIRAs for these TODOs and reference them here (e.g., "TODO(BEAM-1234): ...").

boyuanzz replied:
I added one JIRA in next line. will delete first line in next revision.

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -558,16 +568,33 @@ def __init__(self, window_fn, main_receivers, tagged_receivers):
+    # TODO: [BEAM-3937]
+    experiments = RuntimeValueProvider.get_value('experiments', str, [])
+    if 'outputs_per_element_counter' in experiments:
+      self.has_per_element_output_count = True
```
You can also avoid the `has_per_element_output_count` property by just checking `if self.per_element_output_counter: ...`

boyuanzz replied:
I assumed a boolean variable may be more readable. But I'll get ride of this in next revision.

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -558,16 +568,33 @@ def __init__(self, window_fn, main_receivers, tagged_receivers):
 
+    output_element_output = 0
     for result in results:
+      output_element_output += 1
```
Why not just do `output_element_output = len(results)`?  Also, maybe rename to `output_element_count`?

boyuanzz replied:
ah yes, I'll modify it in next revision. Thanks!

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -546,7 +551,12 @@ def process_outputs(self, windowed_input_element, results):
+               main_receivers,
+               tagged_receivers,
+               counter_factory=None,
+               operation_name=None):
```
Can you please create the counter in `DoFnRunner.__init__` and pass the output counter into this method instead of leaving the logic to create the counter in this method?

Also, please don't make the added argument optional--this file is the only place the `_OutputProcessor` is referenced.

boyuanzz replied:
will fix in next revision

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -449,7 +452,8 @@ def __init__(self,
                state=None,
-               scoped_metrics_container=None):
+               scoped_metrics_container=None,
+               operation_name=None):
```
What is the difference between `step_name` and `operation_name`?

Also, please document any additional arguments in the docstring below

boyuanzz replied:
From my understanding so far, we use operation_name referred to external step(eg. original graph step) and use step_name referred to internal step(eg. optimized graph step). When we initialize a operation, operation.operation_name = operation.step_name, but will change step_name when calling create_operation.

boyuanzz commented:
Hey @charlesccychen ,
I addressed your comments above in latest commit. Please review these changes. Thank you!

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -565,9 +588,14 @@ def process_outputs(self, windowed_input_element, results):
 
+    output_element_count = 0
     for result in results:
+      output_element_count += 1
```
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
Please add comment that we use this increment pattern because `results` may be an iterable.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -405,3 +408,22 @@ def increment_bucket(self, bucket_index):
   _accumulator_type = DistributionAccumulator
+
+
+class DistributionMetadataAccumulator(object):
```
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
Can you please add a class docstring describing how this is to be used?--it is not obvious from reading this file.  This looks like just a generic list accumulator to me.

boyuanzz replied:
Actually, it's just a list. will it be much clearer using another class name such as Int64ListAccumulator?

charlesccychen replied:
Can you add a docstring?

charlesccychen commented on sdks/python/apache_beam/runners/dataflow/internal/apiclient.py:
```diff
@@ -744,6 +744,12 @@ def translate_distribution(distribution_update, metric_update_proto):
   dist_update_proto.max = to_split_int(distribution_update.max)
   dist_update_proto.count = to_split_int(distribution_update.count)
   dist_update_proto.sum = to_split_int(distribution_update.sum)
+  if (hasattr(distribution_update, 'buckets') and
```
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
Why wouldn't the update object have these attributes?

Also, please avoid `hasattr`: https://hynek.me/articles/hasattr/.

boyuanzz replied:
We are using another distribution metric called DistributionUpdate(https://github.com/apache/beam/blob/ce0c92e0a15b9b44e53161827afe213eb18e18b4/sdks/python/apache_beam/metrics/cells.py#L367) for other metrics, which doesn't have buckets and first_bucket_offset

charlesccychen commented on sdks/python/apache_beam/utils/counters.py:
```diff
@@ -134,6 +134,8 @@ class Counter(object):
   # Distribution Accumulator Fn
   DISTRIBUTION = cy_combiners.DistributionCounterFn()
 
+  DISTRIBUTION_METADATA = cy_combiners.DistributionMetadataCounterFn()
```
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
Please add comment; please also add comment indicating this is an experimental API with no backwards compatibility guarantees.

boyuanzz replied:
https://github.com/apache/beam/blob/master/sdks/python/apache_beam/utils/counters.py#L23
We have declared 'no backwards compatibility guarantees'  at the beginning of counter.py

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -565,9 +588,14 @@ def process_outputs(self, windowed_input_element, results):
     then dispatched to the appropriate indexed output.
     """
     if results is None:
+      # TODO[BEAM-3937]:Remove if block after output feature released.
```
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
Use the form `"# TODO(BEAM-1111): My comment"`.

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -463,6 +467,7 @@ def __init__(self,
```
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
Can you clarify this existing comment?  Especially in contrast to `operation_name`.

charlesccychen replied:
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
Can you clarify this existing comment?  Especially in contrast to `operation_name`.

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -477,8 +482,20 @@ def __init__(self,
+    if 'outputs_per_element_counter' in experiments:
+      per_element_output_counter_name = (
+          CounterName('per-element-output-count-meta',
+                      step_name=operation_name))
```
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
Why is the step name the operation name and not the step name?  Please add a clarifying comment for why, if this needs to be the case.

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -477,8 +482,20 @@ def __init__(self,
     # Optimize for the common case.
     main_receivers = tagged_receivers[None]
+
+    # TODO[BEAM-3937]:Remove if block after per-element-output-counter released.
```
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
Use the form `"# TODO(BEAM-1111): My comment"`.

charlesccychen commented on sdks/python/apache_beam/runners/common.pxd:
```diff
@@ -87,15 +87,15 @@ cdef class DoFnRunner(Receiver):
 
 cdef class OutputProcessor(object):
-  @cython.locals(windowed_value=WindowedValue)
+  @cython.locals(windowed_value=WindowedValue, output_element_count=cython.int)
```
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
Please use `int64_t` after importing `from libc.stdint cimport int32_t, int64_t` at the top of the file.

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -592,6 +620,9 @@ def process_outputs(self, windowed_input_element, results):
         self.main_receivers.receive(windowed_value)
       else:
         self.tagged_receivers[tag].receive(windowed_value)
+    # TODO[BEAM-3937]:Remove if block after output feature released.
```
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
Use the form `"# TODO(BEAM-1111): My comment"`.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -405,3 +408,22 @@ def increment_bucket(self, bucket_index):
+    self.distribution_metadata = []
+
+  def add_input(self, element):
+    self.distribution_metadata.append(element)
```
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
It looks like to me (looking at this change and the internal Dataflow worker change) that the reason we need the `DistributionMetadataAccumulator` / `DistributionMetadataCounterFn` classes is that `DistributionCounterFn` is too slow, so here, we add everything to a list, and we do processing for the list later.  Is my understanding correct?

If so, I don't think using this instead of a `DistributionCounter` counter directly changes much--the processing will just get done in the worker API client thread, right, where we create another DistributionCounter called `output_counter` and individually call `output_counter.update(value)` for each value we push onto the list here?  In such a case, we would be hiding the overall overhead of this counter from the processing here, but we don't get rid of it--we just move it to the progress reporting thread?

boyuanzz replied:
Yes, it is. I think, moving distribution calculation to report_progress thread will reduce some overhead overall since report_progress thread doesn't block execution thread, comparing to all distribution calculated in runner code path.

charlesccychen replied:
This is concerning to me because this overhead won't show up in benchmarks of the execution path here, but will be the same in practice--we are just moving it to another thread.  Python execution is essentially single-core-bound because of the Global Interpreter Lock (GIL; http://www.dabeaz.com/python/UnderstandingGIL.pdf).  From the internal Dataflow change, I see that what is being done is equivalent to using a DistributionCounter directly here.  Perhaps we should focus on optimizing the DistributionCounter code path so that this can be used directly here.

CC: @robertwb

boyuanzz replied:
I have one PR which related to DistributionCounter optimization: https://github.com/apache/beam/pull/4928. I'm not sure whether there are other ways to make it better. Do you have some ideas on this?

boyuanzz replied:
As @charlesccychen suggested, I ran map_fn_microbenchmark with DistributionCounter and DistributionMetadataCounter enabled to see whether it's reasonable to update DistributionCounter directly during execution. Results as below:
```
Experimental flag disabled:
     1 element  0.882056 sec
  1001 elements 1.06062 sec
  2001 elements 0.933104 sec
  3001 elements 1.1544 sec
  4001 elements 1.22668 sec
  5001 elements 1.24582 sec
  6001 elements 1.52356 sec
  7001 elements 1.50509 sec
  8001 elements 1.87628 sec
  9001 elements 1.845 sec
Fixed cost   0.824348012299
Per-element  1.11289043138e-06
R^2          0.919787959108
```
```
DistributionMetadataCounter enabled:
     1 element  0.903107 sec
  1001 elements 1.06191 sec
  2001 elements 1.1449 sec
  3001 elements 1.31494 sec
  4001 elements 1.3732 sec
  5001 elements 1.42241 sec
  6001 elements 1.83665 sec
  7001 elements 1.84627 sec
  8001 elements 1.9221 sec
  9001 elements 2.1069 sec
Fixed cost   0.893742784382
Per-element  1.33191660679e-06
R^2          0.969293745853
```
```
DIstributionCounter enabled with optimization(https://github.com/apache/beam/pull/4928)
Distribution enable:
     1 element  0.933321 sec
  1001 elements 1.02975 sec
  2001 elements 1.12517 sec
  3001 elements 1.45851 sec
  4001 elements 1.46911 sec
  5001 elements 1.64176 sec
  6001 elements 1.8858 sec
  7001 elements 2.03899 sec
  8001 elements 2.17605 sec
  9001 elements 2.33426 sec
Fixed cost   0.882123873349
Per-element  1.61552530346e-06
R^2          0.988375238657

```
CC: @robertwb, @pabloem

charlesccychen replied:
Thanks!  @robertwb, can you look at this?  To my understanding, the `DistributionMetadataCounter` and `DistributionCounter` cases above should be equivalent--the `DistributionMetadataCounter` results hide the cost of doing the distribution update in the progress reporting thread, which isn't included in the benchmark.  We should therefore use the `DistributionCounter` numbers in assessing the cost of these changes.

charlesccychen replied:
In non-IO-bound cases, this would contribute in the same way to the total runtime, so putting it in the progress reporting thread is not free.  I will let @robertwb make the decision on the acceptable cost threshold here.

robertwb replied:
While moving computation to the progress reporting thread is not free, there are significant advantages, as the progress reporting thread is called at a much lower frequency (seconds vs. microseconds). This is assuming, of course, that computing a distribution can be amortized across several elements in a manner cheaper than handling each one individually (batched or sequentially). 

It may be an open question whether distributions of the output counts of every single element for every single DoFn can be made cheap enough to enable everywhere; certainly the 20% overhead here is too high. To stay under 5% (using half of a 10% "monitoring budget" for this single feature) we need a distribution structure that can be updated in 50ns. This is certainly possible in C (e.g. imagine a C array that stores count -> bucket for every count up to 1000, and then doing buckets[count_to_bucket[count]] += 1 for another C array buckets, and falling back to something more generic for larger counts (which, naturally, can afford more overhead)).

Not all code is performance critical, but that which runs at every element of every DoFn in every stage in ever pipeline is.

charlesccychen replied:
Right.  My concern is that there is no amortization done right now, if you look at the corresponding proposed changes in the progress reporting code (i.e. there is no `update_multiple()` implementation on `DistributionCounter`), so there may not be much savings when doing this in a batch.

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -546,17 +563,23 @@ def process_outputs(self, windowed_input_element, results):
       window_fn: a windowing function (WindowFn).
       main_receivers: a dict of tag name to Receiver objects.
       tagged_receivers: main receiver object.
+      per_element_output_counter: per_element_output_counter of one work_item
```
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
Please add comment that this can be None to disable this output counter.

boyuanzz commented:
Hey @charlesccychen, latest revision fixed all comments above and python-related tests have passed.

charlesccychen commented:
Thanks Boyuan, this LGTM.

R: @robertwb to review performance and merge.

boyuanzz commented:
Hey @robertwb, could you take a look at latest commit? Charles and Pablo have marked as LGTM. Thanks!

charlesccychen commented:
Hey Robert (@robertwb), can you chime in on the performance issues discussed above?  (https://github.com/apache/beam/pull/4741#discussion_r177570424)  From the discussion, it sounds like you recommended the approach of moving the distribution update to the progress reporting thread, but this approach isn't free and simply hides the overhead from the benchmark.

robertwb commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -565,9 +593,15 @@ def process_outputs(self, windowed_input_element, results):
     then dispatched to the appropriate indexed output.
     """
     if results is None:
+      # TODO(BEAM-3937): Remove if block after output counter released.
```
Add to this TODO to verify performance is not adversely reflected.

boyuanzz replied:
Updated https://issues.apache.org/jira/browse/BEAM-3937

robertwb commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -477,8 +482,24 @@ def __init__(self,
+
+    # TODO(BEAM-3937): Remove if block after output counter released.
+    experiments = RuntimeValueProvider.get_value('experiments', str, [])
+    per_element_output_counter = None
```
Nit: I would put this in an else clause.

boyuanzz replied:
Fixed in latest commit!

boyuanzz commented:
Hey @robertwb, could you please explain more about your count_to_bucket idea? I'm also confused about, how can we combine C array with pure python codes? Thanks so much for your help!

robertwb commented:
One could create a cdef class that has cdef attributes (including C arrays) and a cdef method `update(int)` http://cython.readthedocs.io/en/latest/src/tutorial/cdef_classes.html

boyuanzz commented:
How about count_to_bucket? Are you suggesting, we do distribution update during execution, and use C array to improve performance? @robertwb

robertwb commented:
Yes, that is what I am suggesting.

boyuanzz commented:
Latest commit https://github.com/apache/beam/pull/4741/commits/1af9f1504665185fb37ba9252aa0c4bbe8eba0a3:

Update output_counter with new version of DataflowDistributionCounter. Things changed compared to previous commits:

- Update distribution directly, get rid of changing worker side

- Now we have 2  requests of enabling output_counter: flag_enabled & counter is cythonized

Microbenchmark Performance:
```
Enable output_counter:
     1 element  0.934455 sec
  1001 elements 1.08432 sec
  2001 elements 0.954366 sec
  3001 elements 1.16596 sec
  4001 elements 1.40111 sec
  5001 elements 1.47369 sec
  6001 elements 1.53211 sec
  7001 elements 1.53268 sec
  8001 elements 1.77773 sec
  9001 elements 1.92784 sec
Fixed cost   0.891310163764
Per-element  1.0822390932e-06
R^2          0.940979030477
``` 
```
Disable output_counter:
     1 element  0.913116 sec
  1001 elements 0.986421 sec
  2001 elements 1.15499 sec
  3001 elements 1.31451 sec
  4001 elements 1.28276 sec
  5001 elements 1.42376 sec
  6001 elements 1.50766 sec
  7001 elements 1.70327 sec
  8001 elements 1.76268 sec
  9001 elements 1.82393 sec
Fixed cost   0.921035078612
Per-element  1.03593782945e-06
R^2          0.980515490382
```
Related PR: https://github.com/apache/beam/pull/5028 
@robertwb Please review these new changes. Thanks!

boyuanzz commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -133,3 +133,6 @@ cdef class DataflowDistributionCounter(object):
     """
     return self._fast_calculate_bucket_index(element)
+
+  cpdef merge(self, accumulators):
```
I'm not sure whether it's a good time to implement ```def merge```, I may suggest attaching a JIRA here.

boyuanzz commented on sdks/python/apache_beam/transforms/py_dataflow_distribution_counter.py:
```diff
@@ -112,3 +112,6 @@ def translate_to_histogram(self, histogram):
     histogram.bucketCounts = (
         self.buckets[first_bucket_offset:last_bucket_offset + 1])
+
+  def merge(self, accumulators):
```
Same above

boyuanzz commented:
I feel so sorry I may make you @robertwb @aaltay overloaded, could you please review this PR?

robertwb commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -547,8 +552,24 @@ def __init__(self,
+    # TODO(BEAM-3937): Remove if block after output counter released.
+    experiments = RuntimeValueProvider.get_value('experiments', str, [])
+    # Experimental flag format: experimental-name_version-number.
+    # Specify version number to ensure old SDKs will not be affected by features
```
I'm not quite following this comment, as this code is in the SDK itself.

boyuanzz commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -665,7 +663,7 @@ def process_outputs(self, windowed_input_element, results):
       # TODO(BEAM-3937): Remove if block after output counter released.
       # Only enable per_element_output_counter when counter cythonized.
-      if (self.per_element_output_counter and
+      if (self.per_element_output_counter is not None and
```
Hey @robertwb, could you please explain a little bit here why this caused perf changes? Thanks so much for your help!

robertwb replied:
Sure. 

Writing `if expression:` (or using it in a logical construction with ands/ors) causes `bool(expression_result)` to be called. Generally this is just fine and preferred, but given the frequency this is called we can instead check if it's (not) exactly `None` which compiles down to a single pointer comparison. 

I'll merge this as soon as I verify test (still) pass.

boyuanzz replied:
Thanks so much for your explaination! We are still discussing about how to use experimental flag. We are supposed to reach to an agreement and update this PR for the experimental flag part before this Thursday.

boyuanzz commented:
retest this please

boyuanzz commented:
This PR is ready to merge, but blocked by performance issue.

boyuanzz commented:
Hey @aaltay ,please review the latest commit https://github.com/apache/beam/commit/28dd045fd2ca073a01b1da1bf43281d0d8425419 . Thanks for your help!

boyuanzz commented:
Hey @robertwb, please review the latest 2 commits: 
https://github.com/apache/beam/pull/4741/commits/28dd045fd2ca073a01b1da1bf43281d0d8425419
https://github.com/apache/beam/pull/4741/commits/ade3da2cf98811b1b364f0c799632c47c4283fca

boyuanzz commented:
Revert RuntimeValueProvider related changes. 
Next action:
Wait for https://github.com/apache/beam/pull/5308 merged than apply the same pattern.

boyuanzz commented:
Hey @robertwb, please review the latest commit: https://github.com/apache/beam/pull/4741/commits/f14dbfb4115264e632dfd86cf1a36c3fb11c11de

Related commit: https://github.com/apache/beam/commit/60f90c8dcb229c35a82c7be15e64a89678bae058

charlesccychen commented:
My understanding is that this is deprioritized but still being worked on, so I have removed the "stale" label.

pabloem commented:
This PR had been approved for merge a few weeks back. Developer is busy at
the moment, but we still intend to bring this code in.

On Sun, Jul 8, 2018, 1:43 PM stale[bot] <notifications@github.com> wrote:

> This pull request has been marked as stale due to 60 days of inactivity.
> It will be closed in 1 week if no further activity occurs. If you think
> that’s incorrect or this pull request requires a review, please simply
> write any comment. If closed, you can revive the PR at any time and
> @mention a reviewer or discuss it on the dev@beam.apache.org list. Thank
> you for your contributions.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/4741#issuecomment-403315679>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABPc7Pz7-uFhtriuUiMW5UDoiGfVpj6aks5uEm7tgaJpZM4SRQnS>
> .
>
-- 
Got feedback? go/pabloem-feedback
